tsl/src/backup.c

```diff
@@ -0,0 +1,192 @@
+	DistCmdResult *result_cmd;
+	FuncCallContext *funcctx;
+	XLogRecPtr lsn;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		int name_len = strlen(name);
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (name_len >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errhint("maximum length is %d, while provided name has %d chars",
+							 MAXFNAMELEN - 1,
+							 name_len)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level '%s' is not sufficient for creating a restore point",
+							GetConfigOptionByName("wal_level", NULL, false)),
+					 errhint(
+						 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
+
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (!ts_guc_enable_2pc)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("two-phase commit transactions are not enabled"),
+					 errhint("timescale.enable_2pc must be set to true.")));
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to to create restore point "
+							 "on the data node.")));
+
+		/*
+		 * In order to achieve synchronization across the multinode cluster,
+		 * we must ensure that the restore point created on the access node is
+		 * synchronized with each data node.
+		 *
+		 * We must avoid possible situation when a concurrent remote transaction
+		 * commits on the access node during the creation of the distributed
+		 * restore point and gets into the data nodes WAL before the
+		 * restore point marker written.
+		 *
+		 * To do that we take an exclusive lock on the remote transaction
+		 * table, which will force any concurrent transaction
+		 * wait during their PREPARE phase.
+		 */
+		LockRelationOid(ts_catalog_get()->tables[REMOTE_TXN].id, ExclusiveLock);
+
+		/* To prevent situation when new data node added during the execution */
+		LockRelationOid(ForeignServerRelationId, ExclusiveLock);
+
+		DEBUG_WAITPOINT("create_distributed_restore_point_lock");
+
+		/* create local restore point and on each data node */
+		lsn = XLogRestorePoint(name);
+
+		sql = psprintf("SELECT pg_create_restore_point AS lsn "
+					   "FROM "
+					   "pg_catalog.pg_create_restore_point(%s)",
+					   quote_literal_cstr(name));
+
+		result_cmd = ts_dist_cmd_invoke_on_all_data_nodes(sql);
```

Comment:
Is this correct? It looks like you are saving the result_cmd in the SRF multi-call context, but not in the corresponding memory context.